### PR TITLE
Issue 3279 Add ELMAH Active Scanner

### PR DESCRIPTION
**Pull request by kingthorin**
ElmahScanner - The new scanner.
Messages.properties - Add necessary i18n keys and values.
ascanalpha.html - Add new help entry.
ZapAddOn.xml - Add new change entry, and include the new scanner in the
ascanrules section.

Fixes zaproxy/zaproxy#3279

### CODE REVIEW

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+		URI elmahUri = null;
+		try {
+			elmahUri = new URI(baseUri.getScheme(), null, baseUri.getHost(), baseUri.getPort(), "/elmah.axd");
+		} catch (URIException uEx) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("An error occurred creating a URI for the: " + getName() + " scanner. " + uEx.getMessage(),
+						uEx);
+			}
+		}
+			try {
```
Comment (psiinon) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Mix of spaces and tabs for indenting?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+				newRequest.getRequestHeader().setURI(elmahUri);
+			} catch (URIException uEx) {
+				if (LOG.isDebugEnabled()) {
+				LOG.debug("An error occurred setting the URI for a new request used by: " + getName() + " scanner. "
```
Comment (psiinon) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Ditto

**Comment by kingthorin** [comment]
@psiinon thanks, fixed.

**Comment by kingthorin** [comment]
@thc202?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+import java.io.IOException;
+
+/**
+ * Active scan rule which checks whether or not elmah.axd is exposed.
+ * https://github.com/zaproxy/zaproxy/issues/3279
+ * 
+ * @author kingthorin+owaspzap@gmail.com
+ */
+public class ElmahScanner extends AbstractAppPlugin {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Should be an `AbstractHostPlugin`?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+		newRequest.getRequestHeader().setMethod(HttpRequestHeader.GET);
+		URI baseUri = getBaseMsg().getRequestHeader().getURI();
+		URI elmahUri = null;
+		try {
+			elmahUri = new URI(baseUri.getScheme(), null, baseUri.getHost(), baseUri.getPort(), "/elmah.axd");
+		} catch (URIException uEx) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("An error occurred creating a URI for the: " + getName() + " scanner. " + uEx.getMessage(),
+						uEx);
+			}
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Should return?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		}
+		try {
+			newRequest.getRequestHeader().setURI(elmahUri);
+		} catch (URIException uEx) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("An error occurred setting the URI for a new request used by: " + getName() + " scanner. "
+						+ uEx.getMessage(), uEx);
+			}
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Should return?

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties [hunk 6]
```diff
@@ -19,6 +19,12 @@ ascanalpha.cookieslack.session.destroyed = Dropping this cookie appears to have
 ascanalpha.cookieslack.separator = ,
 ascanalpha.cookieslack.endline = \n
 
+ascanalpha.elmahscanner.name = ELMAH Information Leak
+ascanalpha.elmahscanner.desc = The Error Logging Modules and Handlers (ELMAH [elmah.axd]) HTTP Module was found to be available. This module can leak a significant amount of valuable information.
+ascanalpha.elmahscanner.desc.otherinfo = Based on response status code ELMAH may be protected by an authentication or authorization mechanism.  
+ascanalpha.elmahscanner.refs = https://www.troyhunt.com/aspnet-session-hijacking-with-google/\nhttps://www.nuget.org/packages/elmah\nhttp://elmah.github.io/
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties]: Could use HTTPS for the last one too.

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties [hunk 7]
```diff
@@ -19,6 +19,12 @@ ascanalpha.cookieslack.session.destroyed = Dropping this cookie appears to have
...
+ascanalpha.elmahscanner.soln = Consider whether or not ELMAH is actually required in production, if it isn't then disable it. If it is then ensure access to it requires authentication and authorization. See also: https://code.google.com/archive/p/elmah/wikis/SecuringErrorLogPages.wiki
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/Messages.properties]: Could use the new page: https://elmah.github.io/a/securing-error-log-pages/

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html [hunk 8]
```diff
@@ -16,6 +16,11 @@
 <H2>Cookie Slack Detector</H2>
 Tests cookies to detect if some have no effect on response size when omitted, especially cookies containing the name "session" or "userid"
 
+<H2>ELMAH Scanner</H2>
+Tests to see if the  Error Logging Modules and Handlers (elmah.axd) HTTP Module is available. Although this module is handy for
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/resources/help/contents/ascanalpha.html]: - There's an extra space that could be removed.
 - Worth mentioning the technologies targeted by the scanner.

**Comment by kingthorin** [comment]
Thanks @thc202, I'll tackle that tonight.

**Comment by thc202** [comment]
Thanks!

File: src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		newRequest.getRequestHeader().setMethod(HttpRequestHeader.GET);
+		URI baseUri = getBaseMsg().getRequestHeader().getURI();
+		URI elmahUri = null;
+		try {
+			elmahUri = new URI(baseUri.getScheme(), null, baseUri.getHost(), baseUri.getPort(), "/elmah.axd");
+		} catch (URIException uEx) {
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("An error occurred creating a URI for the: " + getName() + " scanner. " + uEx.getMessage(),
+						uEx);
+				return;
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Ah, just noticed that the return is inside the if not outside (same for following return).

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/ascanrulesAlpha/ElmahScanner.java]: Good catch... will address tonight

**Comment by thc202** [comment]
Thank you!

**Comment by kingthorin** [comment]
@psiinon still good with this?

**Comment by psiinon** [comment]
yeah :)

**Pull request merged**